models/webhook.go

```diff
@@ -798,3 +798,50 @@ func FindRepoUndeliveredHookTasks(repoID int64) ([]*HookTask, error) {
 	}
 	return tasks, nil
 }
+
+// DeleteDeliveredHookTasks deletes delivered hook tasks of one repository, leaving the most recent delivered based on the parameter
+func DeleteDeliveredHookTasks(repoID int64, numberDeliveriesToKeep int64) error {
```

Comment:
Why not just find out last delivery to keep and delete all deliveries in that repository with timestamp lower than that in single delete query. It will be so much faster

Reply:
Good point, that would be simpler logic for sure. However, I thought it would be important to do the deletes in batches to minimize locking on the table. I figured if an installation had been around for a long time, when this goes in it might need to delete an awful lot of rows at least the first time around. Do you think I'm overthinking it? I definitely could change to your suggestion.

Reply:
Single delete query will still be faster than deleting single rows in batches. Also deleting using single queries would be more resource intensive to both Gitea server and database. I would vote for single delete query

Reply:
most guidance I've read recommends breaking the deletes into batches if you have a large dataset, but I'm probably over-estimating how many records there would normally be to delete. I'll change as you recommended, it definitely will be easier to read the logic. thanks!

Reply:
@lafriks see what you think now, I'm going to test it a little more but it seems to work. One query to find the id to base the deletes off of and then the delete statement, no looping.

Reply:
It would be better to use timestamp as later it could be improved to delete hooks older than x days for example

Reply:
I'm making the change as a volunteer programmer for a non-profit organization (unfoldingWord) that uses gitea, and we discussed using a "delete older than x days" concept but "keep x deliveries" worked better for unfoldingWord's use case, so that's the backstory on the implentation. I guess I could still use the delivery date for the delete, but I noticed there's not an index on the delivery time column so I figured using the id would be faster. Any thoughts?

Reply:
I got feedback from the issue to allow the deletes to be done using the "older than x days" method or "keep x deliveries" as implemented here. Probably will be a few weeks before I get to it though.

Reply:
So I think we actually need something like:

```
"DELETE FROM hook_task WHERE id IN (SELECT id FROM hook_task WHERE repo_id = ? AND hook_task.is_delivered = ? ORDER BY hook_task.delivered DESC LIMIT 0, ?", repoID, true, int(numberDeliveriesToKeep)
```

Reply:
I originally tried something similiar but using a subquery on the same table that you are deleting from doesn't work in mysql.

I tried your query as written gives this in mysql:
Error Code: 1235. This version of MySQL doesn't yet support 'LIMIT & IN/ALL/ANY/SOME subquery

but if you take out the limit clause you still get:
Error Code: 1093. You can't specify target table 'hook_task' for update in FROM clause

I read some tricks on SO to get around this but since we need to support sqlite and postgres too I didn't think I should do anything that looks weird.

Reply:
The problem is that although the hook I'd increases monotonically - the delivery time does not necessarily follow that order. It's not hard to imagine a situation whereby id 50 is delivered before 20 but 20 gets deleted in preference to 50.

Reply:
you are right, I updated to delete based on the delivery date